geckocamera-codec.cpp

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright (C) 2021 Open Mobile Platform LLC.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+#include <dlfcn.h>
+#include <map>
+#include <filesystem>
+
+#include "geckocamera-codec.h"
+
+#define LOG_TOPIC "codec"
+#include "geckocamera-plugins.h"
+#include "geckocamera-utils.h"
+
+namespace gecko {
+namespace codec {
+
+using namespace std;
+using namespace gecko::camera;
+
+class RootCodecManager : public CodecManager
+{
+public:
+    RootCodecManager() {};
+    ~RootCodecManager() {};
+
+    bool init();
+
+    bool videoEncoderAvailable(CodecType codecType);
+    bool videoDecoderAvailable(CodecType codecType);
+
+    bool createVideoEncoder(CodecType codecType, shared_ptr<VideoEncoder> &encoder);
+    bool createVideoDecoder(CodecType codecType, shared_ptr<VideoDecoder> &decoder);
+
+private:
+    shared_ptr<CodecManager> loadPlugin(Plugin &plugin);
+
+    bool m_initialized = false;
+    map<const string, shared_ptr<CodecManager>> m_plugins;
+    map<const string, shared_ptr<CodecManager>> m_mimeTypeMap;
+};
+
+bool RootCodecManager::init()
+{
+    if (!m_initialized) {
```

Comment:
Is there any chance multiple threads may call gecko_codec_manager() and concurrently initialize the manager?

Reply:
As far as I understand the initialisation of the decoders is single-threaded, but after a thought I can see that the concurrency can come from webrtc code.